Editing toasts flashes #1385

Jsmi8334 opened the pull request:
## Issue Description

Issue: #1385
All over CELTS different pages use different kind of flash messages. We have toasts, flash messages from flask, and flash messages from Javascript. They all behave differently and look differently. It would be a good idea to attempt to normalize these the best we can.

## Changes

The logic we used for implementing flashes and toasts and custom validity tags is that we are using toasts when we are on a Modal and we need updates, success, deletion messages popped up next to the modal. If there is a required field a custom Validity tag will show up.  We are using flashes for page reloads and anything that does not use a modal so the user will see the flash in the center of the page. We are using Custom validity tags for all required fields throughout the page.

this is a toast: 
![image](https://github.com/user-attachments/assets/5f3aaf04-35b5-4bce-8f5f-b3370c61d537)

This is a flash: 
![image](https://github.com/user-attachments/assets/3757ae61-5f51-4adf-8ae4-fa6c0910ad69)

This is a custom validity tag:
![image](https://github.com/user-attachments/assets/8dd5d921-db34-41da-b20d-1d3abbec3b25)


- Create Event page we added custom validity tags to all required fields. On the same page we changed the order of the (Must select one) and the optional boxes so the user can easily recognize the required fields 
- On my profile we added success, deletion messages for adding background checks, as well as for the notes section
- On my profile we created flashes for the changing of the phone number 
- Fixed all toast to be flash messages and added custom validity tag to Status and Date in Description
- On the proposal For Service-Learning Course page we created custom validity tags for the required fields.
- On Settings --> program management page --> edit manager, we added msg toast for Manager deletion. 
- On my profile --> Description, we added the required sign onto the status and date of background check 
- We added a timer to the toast 3 sec (you can test this by creating a note under My Profile --> Note --> save)

## Testing

- User Profile: When opening Background Check the status and the date are required fields and they now show custom validity tags. If bg check is deleted then we get flash message.
![image](https://github.com/user-attachments/assets/bc17d4ac-9765-43d8-aacb-2cd17c85e21a)
- Create Event pg: 
- All of the required fields will show a custom validity tag if not filled out. (asteris- shows required field). There are three main forms a user can fill out are All Volunteer Training, Bonnar scholar, and the rest of the other avaible forms. 
![image](https://github.com/user-attachments/assets/b85c5ce3-186a-4d20-b61a-f332eb5cb5c2)
- we added success and deletion messages for all of the specific pages in itself for the three main pages shown above. When a form is submitted or deleted or if there is an error. 
- we edited the series of events modal if the toggle is switched on:  Adding custom validity tags to the required fields. This functionality is also seen when on the modal and "this series repeats weekly is selected. 
- Under Admin settings: 
![image](https://github.com/user-attachments/assets/e636e851-1702-47b6-931c-eed29312c611)
-  if the removed button is pressed then a success flash will show up for deletion
- On the same page and under program management if Edit details or Edit managers is selected a success flash will show. On the Edit Managers Modal, if a manager is removed a toast will pop up.
![image](https://github.com/user-attachments/assets/b6e2fd5d-7cad-482d-b62b-8e09a56f1dca)
- On the CCE Minor Page 
![image](https://github.com/user-attachments/assets/c810d945-5fb5-4f59-9a42-b9bc5435e9a5)  if declare student button is pressed or you remove a student we added a Flash
![image](https://github.com/user-attachments/assets/68ce7eab-7871-4dbe-aac0-fd9452d345ae)

Jsmi8334 committed:
Merge branch 'development' of https://github.com/BCStudentSoftwareDevTeam/celts into verbiage

Jsmi8334 committed:
created a reportValidity(). to the create event page to create a pop up that matches the layout of the page better

Jsmi8334 committed:
We got all of the custom validity Tags to show up and work correctly. For all create event templates.

Jsmi8334 committed:
We fixed the submission button  for all different create event pages, we also fixed the custom validity  pop ups on all the create event pages, and we also fixed the layout for for the event and options object on the create the event page

Jsmi8334 committed:
Merge branch 'development' of https://github.com/BCStudentSoftwareDevTeam/celts into Editing-Toasts-Flashes-#1385

Jsmi8334 committed:
fixed all toast to be flash messages and added custom validity tag to Status and Date in Description

Jsmi8334 committed:
created custom validity tags for the required fields on Proposal For Service-Learning Course page

Jsmi8334 committed:
added msg toast for Manager deletion on Settings --> program management page --> edit manager

Jsmi8334 committed:
We created flashes for the removal and addition of Bonner requirements.

Jsmi8334 committed:
Merge branch 'development' of https://github.com/BCStudentSoftwareDevTeam/celts into Editing-Toasts-Flashes-#1385

Jsmi8334 committed:
reverting a commit that would cause a conflict with Javier and Zac

Jsmi8334 committed:
we deleted one of tthe flash messages from the notes section of the user Profile. we also added a succsess note onto the user prfil section

Arohasina committed:
Merge branch 'development' of https://github.com/BCStudentSoftwareDevTeam/celts into Editing-Toasts-Flashes-#1385

Arohasina committed:
Merge branch 'Editing-Toasts-Flashes-#1385' of https://github.com/BCStudentSoftwareDevTeam/celts into Editing-Toasts-Flashes-#1385

Jsmi8334 committed:
We added the required sign to the date and status

Jsmi8334 committed:
Merge branch 'Editing-Toasts-Flashes-#1385' of https://github.com/BCStudentSoftwareDevTeam/celts into Editing-Toasts-Flashes-#1385

BrianRamsay commented on app/static/js/base.js:
```diff
@@ -114,7 +114,7 @@ function validatePhoneNumber(editButtonId, phoneInputId, username) {
       success: function(s){
         $(phoneInputId).attr("data-value",phoneInput.val())
-        msgToast("Phone Number", "Successfully updated the phone number.")
+        msgFlash("Successfully updated the phone number.","success" )
```
Aren't we wanting to use toasts for ajax calls?

Jsmi8334 replied:
we thought it looked more normalized as a flash whenever the phone number is updated since we decided to use toast only for modals. 
but if you like the style better as a toast then we can switch it to a toast

BrianRamsay commented on app/static/js/bonnerManagement.js:
```diff
@@ -76,6 +76,9 @@ $(document).ready(function(){
         }
+        // const fileName = noOfYears === "all" 
+        //     ? "Bonner Spreadsheet, All Cohorts" 
+        //     : `Bonner Spreadsheet, ${Number(startingYear) - Number(noOfYears) + 1} - ${Number(startingYear) + 1}`
```
What's up with this commented out section? It seems like it makes more sense than the if statement above it. Was it causing an error having fileName redefined?

BrianRamsay reviewed (changes requested):
There are a lot of debug comments in your javascript. Also, let's be consistent with using jquery to select elements. 

Particularly for the create events validation function, it would be helpful to have some comments that indicate what different sections of code are doing.

Jsmi8334 committed:
Removed the commented section

Jsmi8334 committed:
added timer to toast

Jsmi8334 committed:
removed debug console.log and added comments to functions and logic

Karina-Agliullova commented:
When I create or edit an event, there is a modal for selecting a series of events. I can create event on different dates and time with different names in that modal. As of now, it doesn't update me when I create or delete events on that modal. Would be nice to have a flsha message for both occasions (addition and deletion) once details in the modal are saved and page is reloaded to update the user that the change has been made

Karina-Agliullova commented on app/static/js/userProfile.js:
```diff
@@ -265,16 +267,20 @@ $(document).ready(function(){
-          bgStatusInput.addClass("invalid");
-          window.setTimeout(() => bgStatusInput.removeClass("invalid"), 1000);
+           $('.form-select').each(function() {
+                bgStatusInput[0].setCustomValidity("Please enter a status");
```
The validity tag shows "Please enter status" even when the status is already filled. The validity tag should disappear when the person selects the status.

Karina-Agliullova commented on app/static/js/createEvents.js:
```diff
@@ -384,6 +384,120 @@ function formatDate(originalDate) {
+        return; // Skip these, they'll be validated separately
+      }
+
+      if ($(this).val() === ""){
```
The value should be trimmed if the person uses empty space - .trim()

Karina-Agliullova commented on app/static/js/createEvents.js:
```diff
@@ -384,6 +384,120 @@ function formatDate(originalDate) {
+      } 
+      else if (isAllVolunteer) {
+      $(".allV").each(function(){
+        if (!$(this).is(":visible") || $(this).is(":disabled")) return;
```
The validity tag should disappear when the value is actively being entered in the datepicker, the event name, the description, the location, and the toggles. It may also be on one of the lines below, possibly, reportValidity, line 454. For example, once they start entering the description, the tag should disappear.

Jsmi8334 committed:
on Create your event --> series of events --> had a flash for required fields i changed the mesage to cusom validity tags. also updates the jquery

Jsmi8334 committed:
Added .trim() to all of the fucntions that checck if the value is empty.

Karina-Agliullova committed:
Merge branch 'development' into Editing-Toasts-Flashes-#1385

Karina-Agliullova commented on app/static/js/createEvents.js:
```diff
@@ -167,19 +167,27 @@ $('#saveSeries').on('click', function() {
     if (eventNameInput.value.trim() === '') {
       isEmpty = true;
-      $(eventNameInput).addClass('border-red');
+      //$(eventNameInput).addClass('border-red');
```
Remove the comment lines that are not used in the code in this file

Karina-Agliullova commented on app/static/js/minorAdminPage.js:
```diff
@@ -19,6 +19,7 @@ $(document).ready(function() {
           contentType: "application/json",
           success: function(response) {
             location.reload()
+            msgFlash("Candidate minor succsessfully removed", "success")
```
This message disappears very quickly after a page reload when i remove a student from the table. Make it stay for longer

Karina-Agliullova replied:
At times, theflash message doesn't even appear on the screen when I remove the student

Jsmi8334 committed:
-fix(event-form): validate fields and fix submission issue on series/single event creation
-fix: added toasts and flash messages for deletion, edit, save on the series modals
-feat(validation): add centralized checkValidation function to handle required field logic
feat(validation): implement live custom validity clearing on input for all event modes
-fix(series): prevent submission if repeatingEventsStartDate >= repeatingEventsEndDate
-fix(series): added display message when start time > end time
-feat(ui): add border-red class and custom error messages to empty or invalid fields
-feat(modal): preserve input state when reopening series modal via edit butt

Jsmi8334 committed:
Merge branch 'Editing-Toasts-Flashes-#1385' of https://github.com/BCStudentSoftwareDevTeam/celts into Editing-Toasts-Flashes-#1385

Jsmi8334 committed:
removed spaces and comments

Jsmi8334 committed:
we fixed the timing of the flash on the Minor Admin Page. When a student is removed

bakobagassas commented on app/static/js/userProfile.js:
```diff
@@ -265,16 +267,20 @@ $(document).ready(function(){
           bgStatusInput.focus()
-          bgStatusInput.addClass("invalid");
-          window.setTimeout(() => bgStatusInput.removeClass("invalid"), 1000);
+           $('.form-select').each(function() {
```
when we save either the date or the status in the background check (one of them being empty and the other one being completed) twice, then we come back to the row, there is a custom validity tag that remains stuck to the status box even after we select something from the status box until we click somewhere on the screen and then the custom validity tag finally disappears (though I don't know if that is actually a problem)

bakobagassas commented on app/static/js/createEvents.js:
```diff
@@ -148,103 +148,139 @@ function createOfferingModalRow({eventName=null, eventDate=null, startTime=null,
+    datePickerInputs.each((index, datePickerInput) => {
+      if (datePickerInput.value.trim() === '') {
+        hasErrors = true;
+        $(datePickerInput)[0].setCustomValidity("Please enter an event date");
```
When we try to save a fully empty form, there is a custom validity tag that appears for the "must select one" section, while the calendar displays at the same time. However, we should have the tag appearing at the calendar's level instead, since the calendar is the one opening up at first.

Jsmi8334 committed:
fixed clearing the custom validity tag on user profile --> background check for status and date when saved

Karina-Agliullova committed:
Merge branch 'development' into Editing-Toasts-Flashes-#1385

Karina-Agliullova commented:
I was testing creating all volunteer training, and i discovered that if I create series of events, and try to delete it in that modal, there is a toast message with Deletion info saying that "You have successfully deleted a serie of events." There is a typo, should be saying "series" instead of "serie." Very minor fix. I can't find that message in your code though, maybe that toast message was created before you picked up the issue, but would be nice if you could fix it

Arohasina committed:
Merge branch 'development' of https://github.com/BCStudentSoftwareDevTeam/celts into Editing-Toasts-Flashes-#1385

Jsmi8334 committed:
merging into devlopment

Karina-Agliullova commented:
When I add a note on User Profile and hit save button, it save and toast message is displayed for less than a second, it feels like. I can't even tell what is written on it. Make the toast message for Notes longer

Jsmi8334 committed:
Merge branch 'Editing-Toasts-Flashes-#1385' of https://github.com/BCStudentSoftwareDevTeam/celts into Editing-Toasts-Flashes-#1385

Arohasina committed:
Merge branch 'Editing-Toasts-Flashes-#1385' of https://github.com/BCStudentSoftwareDevTeam/celts into Editing-Toasts-Flashes-#1385

Karina-Agliullova commented:
Can we add Validation Tags for text and datepicker for Ban/Unban Volunteer modal?
![Screenshot (106)](https://github.com/user-attachments/assets/0192ec3d-d0c1-488d-a5a6-f7a950903a19)

edit: Brian told us we should leave it for now

Karina-Agliullova commented:
Can we add flash message when cancelling event? As of now, there is nothing showing up when I cancel event except that text under the title that stays there for as long as the event is cancelled.
![Screenshot (107)](https://github.com/user-attachments/assets/d3aa8155-f6d4-4a20-a4dc-7b9532ffcbff)

Jsmi8334 committed:
changed the wording of the msgToast on create Event page

Karina-Agliullova commented:
Weird behavior of validity tags when creating course proposals. When I create a new course proposal and it asks me for course instructor, course name, etc. for some reason, if I click save and continue, it doesn't let me and shows me validity tag, which is good. However, that validity tage points at Course instructor, which is the first question asked in that form, but it puts my cursor for typing something into a Course name text box, so it seems like that validity tag is now associated with a different text box
![Screenshot (108)](https://github.com/user-attachments/assets/c305441b-8e5b-433a-8e8d-e65caf14dd1f)

Jsmi8334 committed:
Merge branch 'Editing-Toasts-Flashes-#1385' of https://github.com/BCStudentSoftwareDevTeam/celts into Editing-Toasts-Flashes-#1385

Karina-Agliullova commented:
Related to creating course proposals. There is a behavior similar to the one that you fixed on another page. It asks me with validity tag to input course isntructor, and when I do, it still shows me a validity tag for that specific textbox, even though I added one instructor already

Arohasina committed:
Fix toast duration

Arohasina committed:
refactor: consolidate field validation logic into reusable functions

Arohasina committed:
cleaned unused code

Arohasina committed:
added a flash msg when event is cancelled

Arohasina committed:
Fix custom validity messages appearing on wrong form elements

Arohasina committed:
automatically clears custom validation error messages when users start typing

Arohasina committed:
added flash message when background check is added

Arohasina committed:
fixed "Candidate minor succsessfully removed" to appear after page reloads

Arohasina committed:
fxed startDatePicker to look like the default date picker that we have on other fields. This also fixed the custom validity tag that appears while the calendar displays at the same time.

Arohasina committed:
removed commented codes

Arohasina committed:
Merge branch 'development' of https://github.com/BCStudentSoftwareDevTeam/celts into Editing-Toasts-Flashes-#1385

BrianRamsay commented on app/static/js/minorAdminPage.js:
```diff
@@ -2,29 +2,43 @@ import searchUser from './searchUser.js'
+            sessionStorage.setItem('flashMessage', JSON.stringify({
+                message: "Candidate minor successfully removed",
+                type: "success"
+            }));
```
We should consolidate the session stuff in our `msgFlash` function, so that people who want to use flashes don't have to know how they are persisted. Maybe you could just call `msgFlash` like usual, but with a parameter that indicates it should be after the next page load, instead of immediately.

Arohasina committed:
Create Events Page -->  Volunteer Event Form we  fixed the Save button location

Arohasina committed:
Merge branch 'development' of https://github.com/BCStudentSoftwareDevTeam/celts into Editing-Toasts-Flashes-#1385

BrianRamsay commented on app/static/js/userProfile.js:
```diff
@@ -216,8 +216,13 @@ $(document).ready(function(){
+            setTimeout(function() {
+                reloadWithAccordion(target)
+                location.reload()
+            }, 2500) 
```
Shouldn't this be using the new flash that shows up on the next page reload instead of a toast?

BrianRamsay commented on app/static/js/userProfile.js:
```diff
@@ -239,7 +245,24 @@ $(document).ready(function(){
+      }
+    });
+  });
+}
```
We create this function twice on different pages. We should probably have it in base.js and you can just pass in the selectors you want to work with.

BrianRamsay commented on app/static/js/userProfile.js:
```diff
@@ -249,41 +272,48 @@ $(document).ready(function(){
+            msgFlash(`Successfully added background check`, "success", 1300)
+            setTimeout(function() {
+            reloadWithAccordion("background")
+            }, 1500)
```
Same detail here, instead of delaying reload we could just show the flash after reload

BrianRamsay commented on app/templates/events/createEvent.html:
```diff
@@ -75,14 +75,10 @@ <h1 id="pageTitle">{{page_title}}</h1>
-      <input autocomplete="off" type='text' class="form-control datePicker startDatePicker readonly"
-        value="{{ eventData.startDate.strftime('%Y/%m/%d') if eventData.startDate and eventData.startDate.strftime else eventData.startDate}}"
-        name="startDate" placeholder="Pick a start date" id='startDatePicker-{{pageLocation}}'
+      <input autocomplete="off" type='date' class="form-control datePicker startDatePicker readonly main allV"
```
There are a few places with `all` and `allV` classes, but I don't see anywhere they are referenced in css or js. What do they do?

Arohasina replied:
it is a way to classify the different required fields based on what type of template they are on
allv: start date and event name (on the modal)
all: location, description, event name 
we had to split them up so that only the field that are on the page are being checked and it would allow submission

BrianRamsay commented on app/templates/events/createEvent.html:
```diff
@@ -246,63 +241,58 @@ <h1 id="pageTitle">{{page_title}}</h1>
+{% endif %}
+<div class="form-group mb-5">
+  <label class="mb-2" for='attachmentObject'><strong>Add Files</strong></label>
+  <input type="file" class="form-control" id="attachmentObject" name="attachmentObject" multiple />
```
Looks like we lost the list of file types it accepts (old line 304). Was that intentional?

Arohasina replied:
we added it back

BrianRamsay commented on app/templates/events/createEvent.html:
```diff
@@ -466,22 +456,21 @@ <h5 class="modal-title">This is an event with multiple time offerings.</h5>
               <div class='input-group date'>
                 <input autocomplete="off" type='date'
-                  class="form-control repeatingEventsEndDate repeatingEventsField readonly"
+                  class="form-control repeatingEventsEndDate repeatingEventsField readonly seriesWeekly"
```
Is `seriesWeekly` used anywhere?

Arohasina replied:
yes in this function:
function checkValidation() in createEvents.js line 452

BrianRamsay commented on app/templates/events/createEvent.html:
```diff
@@ -551,7 +540,7 @@ <h3 class="mt-1 d-flex justify-content-center"> Generated Events: </h3>
               data-page-location="{{pageLocation}}">
               <input autocomplete="off" type='date'
-                class="form-control datePicker multipleOfferingDatePicker readonly"
+                class="form-control datePicker multipleOfferingDatePicker readonly-series"
```
Is `readonly-series` used anywhere?

Arohasina replied:
should be series

BrianRamsay commented:
As part of these changes we should write up a page in http://ssdt-documentation.berea.edu about our usage of flashes and toasts, and what our policy is, and a page about using custom validity, what we want, and how to implement it.

Arohasina committed:
consolidate sessionStorage logic inside msgFlash

Arohasina committed:
We added a message flash to the student intresst  page  when  a student is added --> minor managment form

Arohasina committed:
got rid of a space

Arohasina committed:
consolidate flash message loading logic inside msgFlash function

Arohasina committed:
used msgflash that shows up on the next page reload instead of a toast

Arohasina committed:
reverted id changes for must select toggles

Arohasina committed:
refactor enableLiveCustomValidityClearing() to base js and changesd the calls to the function

Arohasina committed:
delayed page reload using new msgFlash() function

Arohasina committed:
fixed syntax error

Arohasina committed:
Slc --> course proposal --> we added a flash message for the removal of the instructor

Karina-Agliullova reviewed (approved):
Everything looks good!

BrianRamsay committed:
Merge branch 'development' into Editing-Toasts-Flashes-#1385

Arohasina commented:
added Flashes, Toasts, and Custom Validity Tags explanation and usage to the documentation: http://ssdt-documentation.berea.edu/en/ssdt-program/FlashesToasts
